services/notifier/notifier.go

```diff
@@ -0,0 +1,588 @@
+		"workspace": n.workspaceIdentifier,
+		"module":    "pg_notifier",
+		"queueName": queueName,
+	})
+	return n
+}
+
+func (n *Notifier) Setup(
+	ctx context.Context,
+	fallbackDSN string,
+) error {
+	dsn := fallbackDSN
+	if n.checkForNotifierEnvVars() {
+		dsn = n.connectionString()
+	}
+
+	if err := n.setupDatabase(ctx, dsn); err != nil {
+		return fmt.Errorf("could not setup db: %w", err)
+	}
+	n.repo = newRepo(n.db)
+
+	groupCtx, groupCancel := context.WithCancel(ctx)
+	n.background.group, n.background.groupCtx = errgroup.WithContext(groupCtx)
+	n.background.groupCancel = groupCancel
+	n.background.groupWait = n.background.group.Wait
+
+	return nil
+}
+
+func (n *Notifier) checkForNotifierEnvVars() bool {
+	return n.conf.IsSet("PGNOTIFIER_DB_HOST") &&
+		n.conf.IsSet("PGNOTIFIER_DB_USER") &&
+		n.conf.IsSet("PGNOTIFIER_DB_NAME") &&
+		n.conf.IsSet("PGNOTIFIER_DB_PASSWORD")
+}
+
+func (n *Notifier) connectionString() string {
+	return fmt.Sprintf("host=%s port=%d user=%s password=%s dbname=%s sslmode=%s",
+		n.config.host,
+		n.config.port,
+		n.config.user,
+		n.config.password,
+		n.config.database,
+		n.config.sslMode,
+	)
+}
+
+func (n *Notifier) setupDatabase(
+	ctx context.Context,
+	dsn string,
+) error {
+	database, err := sql.Open("postgres", dsn)
+	if err != nil {
+		return fmt.Errorf("could not open: %w", err)
+	}
+	database.SetMaxOpenConns(n.config.maxOpenConnections)
+
+	if err := database.PingContext(ctx); err != nil {
+		return fmt.Errorf("could not ping: %w", err)
+	}
+
+	n.db = sqlmw.New(
+		database,
+		sqlmw.WithLogger(n.logger.Child("notifier-db")),
+		sqlmw.WithQueryTimeout(n.config.queryTimeout),
+		sqlmw.WithStats(n.statsFactory),
+	)
+
+	if err := n.setupTables(); err != nil {
+		return fmt.Errorf("could not setup tables: %w", err)
+	}
+	return nil
+}
+
+func (n *Notifier) setupTables() error {
+	m := &migrator.Migrator{
+		Handle:                     n.db.DB,
+		MigrationsTable:            "pg_notifier_queue_migrations",
+		ShouldForceSetLowerVersion: n.config.shouldForceSetLowerVersion,
+	}
+
+	operation := func() error {
+		return m.Migrate("pg_notifier_queue")
+	}
+
+	backoffWithMaxRetry := backoff.WithMaxRetries(backoff.NewExponentialBackOff(), 3)
+
+	err := backoff.RetryNotify(operation, backoffWithMaxRetry, func(err error, t time.Duration) {
+		n.logger.Warnf("retrying warehouse database migration in %s: %v", t, err)
+	})
+	if err != nil {
+		return fmt.Errorf("could not migrate pg_notifier_queue: %w", err)
+	}
+	return nil
+}
+
+// ClearJobs deletes all jobs for the current workspace.
+func (n *Notifier) ClearJobs(ctx context.Context) error {
+	if n.workspaceIdentifier == "" {
+		return nil
+	}
+
+	n.logger.Infof("Deleting all jobs for workspace: %s", n.workspaceIdentifier)
+
+	err := n.repo.resetForWorkspace(ctx, n.workspaceIdentifier)
+	if err != nil {
+		return fmt.Errorf("could not reset notifier for workspace: %s: %w", n.workspaceIdentifier, err)
```

Comment:
We're missing test cases for the repo methods returning errors as well.

Reply:
Added tests for these.